BuildConfiguration.get() now returns string type value for string options, instead of array of sub-strings

de-nordic opened the pull request:
The PR contains two changes:
1) changes the behavior of the `BuildConfiguration.get()` method that now returns strings instead of lists of sub-strings;
the real change has been done to `_parse_value()` method, which used to split found string into list of sub-strings, with a white-space being the split point, but the visibility is at the level of `BuildConfiguration.get()`;
2) ImgtoolSigner.sign() method has been altered to use `BuildConfiguration.get()` to obtain Kconfig options in places where `load_dot_config` obtained values have been used; this allowed to remove `load_dot_config` from the method and as a result prevented `.config` file from being loaded and parsed second time.

mbolivar-nordic commented on scripts/west_commands/runners/core.py:
```diff
@@ -153,8 +158,9 @@ def _parse(self, filename: str):
+        # If it is a quoted string, return it with quotation marks removed
         if value.startswith('"') or value.startswith("'"):
-            return value.split()
+            return value[1:-1].strip()
```
I think it needs to be stripped first:

```
value.strip()[1:-1]
```

Otherwise, a manually edited file with whitespace at the end of a string option will fail to parse as intended:

```
CONFIG_MY_STRING="foo"        
                      ^^^^
                      manual whitespace
```

de-nordic replied:
> 
> 
> I think it needs to be stripped first:
> 
> ```
> value.strip()[1:-1]
> ```
> 
> Otherwise, a manually edited file with whitespace at the end of a string option will fail to parse as intended:
> 
> ```
> CONFIG_MY_STRING="foo"        
>                       ^^^^
>                       manual whitespace
> ```

Hm... I want to keep my `strip`, lets to:
```
value.strip()[1:-1].strip()
```
Every white-space gets stripped.

mbolivar-nordic replied:
Can you explain why you want to strip whitespace that the user included explicitly?

de-nordic replied:
Good point. Somehow I thought that leading spaces in " stripped string are also bad, it did not occurred to me that it can be intentional. Will fix.

tejlmand replied:
> Can you explain why you want to strip whitespace that the user included explicitly?

actually, existing code did do a strip, so maybe we should keep that, see also: https://github.com/zephyrproject-rtos/zephyr/pull/30214#discussion_r604750095

Especially `CONFIG_TEST="My string"    ` (<-- notice trailing whitespaces) will fail, as the `"` is no longer matched on `value.endswith('"')`

mbolivar-nordic replied:
@tejlmand if I understand you, you're suggesting we do this instead:

```py
                    value = match.group('value').rstrip()                              # notice: rstrip(), NOT strip()
                    if value.startswith('"') and value.endswith('"'):
                        value = value[1:-1]
```

Instead of the current `value = match.group('value')`. That would allow _trailing_ whitespace in the value, which is consistent with what `BuildConfiguration._parse` does today, but is not consistent with `load_dot_config()`. We have to pick one behavior, though, and I think allowing trailing whitespace like `BuildConfiguration` does makes sense, especially given how sloppy a lot of modern editors are when it comes to inserting it.

The current code in both `load_dot_config()` and `BuildConfiguration._parse` does _not_ strip _leading_ whitespace from the value though, so `strip()` would not be consistent with either one. That is, `CONFIG_FOO=     y` is not treated the same as `CONFIG_FOO=y` by either function, and I don't think we should make a behavior change here.

tejlmand replied:
> 

> @tejlmand if I understand you, you're suggesting we do this instead:

Not necessarily there, but that would be a fine place to do it.
I was merely observing that the regex would include trailing whitespaces, but that didn't happen with old code, that looked like:
```
line = line.strip()
if not line or line.startswith('#'):
    continue
option, value = line.split('=', 1)
```

Of course there can be good reasons to change old behavior, but as you so nicely put it:
> especially given how sloppy a lot of modern editors are when it comes to inserting it.

then I think the `value = match.group('value').rstrip()` proposal looks very nice.
I rarely think users places whitespaces in front of settings, like: `   CONFIG_TEST=y`.
More common to get unintended trailing whitespaces.

Note: whitespaces in front of values has never been stripped, so I don't think we should start doing that.

mbolivar-nordic replied:
OK, sounds like we are on the same page and the requested change is to use:

```py
                    value = match.group('value').rstrip()
```

Thanks for catching this.

mbolivar-nordic commented on scripts/west_commands/runners/core.py:
```diff
@@ -139,8 +140,12 @@ def __getitem__(self, item):
     def get(self, option, *args):
         return self.options.get(option, *args)
 
+    def path(self):
```
Just a minor thing, but could you initialize an attribute named `path` in the `__init__` method directly, to avoid adding this method?

Something like this in `__init__`:

```py
self.path = os.path.join(self.build_dir, 'zephyr', '.config')
self._parse(self.path)
```

de-nordic replied:
> 
> Just a minor thing, but could you initialize an attribute named `path` in the `__init__` method directly, to avoid adding this method?
> 

It would be possible though I want to keep the `path` method as I wanted to somehow, implicitly, suggest that path can only be obtained and is set internally.

mbolivar-nordic replied:
In Python there is no need for that -- the implementation details can be changed from an attribute to a property later if necessary without users having to change their code.

de-nordic replied:
OK, will fix. I am too much from the C++/Java ground where you allow attribute access via getters/setters rather than directly.

mbolivar-nordic commented on scripts/west_commands/sign.py:
```diff
@@ -18,8 +18,7 @@
 from zcmake import CMakeCache
-from zephyr_ext_common import Forceable, load_dot_config, \
-    ZEPHYR_SCRIPTS
+from zephyr_ext_common import Forceable, ZEPHYR_SCRIPTS
```
Ah, I seem to have forgotten about BuildConfiguration when I wrote `load_dot_config()`.

If you check the implementation of `load_dot_config()`, it is actually more robust than `BuildConfiguration._parse`. For example, `BuildConfiguration` does not understand that `# CONFIG_FOO is not set` really means `CONFIG_FOO=n`.

So making this change is a bit of a step backwards in terms of compatibility with kconfiglib.

I suggest that what we do is move the code for `load_dot_config` into `BuildConfiguration._parse`, since `load_dot_config` is actually doing the right thing (even for substrings).

de-nordic replied:
> 
> If you check the implementation of `load_dot_config()`, it is actually more robust than `BuildConfiguration._parse`. For example, `BuildConfiguration` does not understand that `# CONFIG_FOO is not set` really means `CONFIG_FOO=n`.
> 
Yes I know I have seen this, but the `BuildConfiguration` is used more often...

> So making this change is a bit of a step backwards in terms of compatibility with kconfiglib.
> 
> I suggest that what we do is move the code for `load_dot_config` into `BuildConfiguration._parse`, since `load_dot_config` is actually doing the right thing (even for substrings).

.. and yes I was going to propose changing that next, though as it does not seem to be that crucial: the `get(key, alternative_value)` fills the gap

mbolivar-nordic replied:
> Yes I know I have seen this, but the `BuildConfiguration` is used more often...

Right, that's why I think we can just move load_dot_config's code into BuildConfiguration's _parse method, right?

de-nordic replied:
OK, I will do that.

de-nordic replied:
Ok, so I have partially merged the `load_dot_config` into the _parse().
I have left out generating dictionary entries for the "unset values", for following reasons:
1) The `n` is not the same as unset because unset value may be also number or string, which means that in case where string or number is unset it being actually set to 'n';
2) not adding option to directory allows to detect that the option is really unset rather set to `n`
3) The default value parameter to `get` works better than defaulting all unset options to `n` because user does know what is the type of an option that is being checked with `get` and can decide what should the default be, or can detect if value is unset;
4) It is not true that .config is not always automatically generated and contains strict syntax: for example I tend to edit parameters directly in .config; this means that some lines get commented out by me and are not ending wit "not set" and fail to be properly matched.

mbolivar-nordic commented on scripts/west_commands/runners/core.py:
```diff
@@ -140,25 +142,41 @@ def get(self, option, *args):
+        self._parse(self.path)
 
     def _parse(self, filename: str):
+        parser = re.compile("^(?P<prefix>#*)[ \t]*(?P<option>CONFIG_[a-zA-Z0-9_]+)=(?P<value>.*)")
```
Thanks for the update!

I can't find anything in this PR that preserves the explicit match for `# CONFIG_FOO is not set` -> `CONFIG_FOO=n`, though; am I missing it?

de-nordic replied:
I have argued why the `is not set` match does not add value here https://github.com/zephyrproject-rtos/zephyr/pull/30214#discussion_r531233454, but If you really want it I can match the `is not set` lines to `None`.

mbolivar-nordic reviewed (commented):
As we discussed offline, I am requesting that the logic in load_dot_config be copy/pasted into BuildConfiguration without changing the semantics.

The purpose of this is to load `.config`. Users are free to edit prj.conf by hand, but you shouldn't edit `.config` by hand unless you "know what you're doing" and can match the kconfiglib output carefully.

The current PR is changing the semantics where `# CONFIG_FOO is not set` lines are concerned. The regular expression logic is different.

de-nordic commented:
> 
> As we discussed offline, I am requesting that the logic in load_dot_config be copy/pasted into BuildConfiguration without changing the semantics.

Yes, I have responded `OK, I will fix it.`

> 
> The current PR is changing the semantics where `# CONFIG_FOO is not set` lines are concerned. The regular expression logic is different.

No it does not. Yes the logic is different, the result is the same.

The outcome of this code is exactly the same dictionary object as in case of the `load_dot_config`, all the `is not set` options.
The only difference is in console output: there will be warning if there is 'CONFIG_something=' with no value present, which can be discarded from code and then both outcomes, dictionary and console, are the same.

In contrary to previous code, the new regular expressions catches both cases in one run; the previous code would need to run second RE on an input line if the first RE would not match: here everything happens in a single RE run and is then processed by logic.

The cited source of REs: https://github.com/torvalds/linux/blob/master/scripts/kconfig/merge_config.sh#L107
does not actually define REs but sed expressions for extraction of Kconfig option names from .config file, which means that from this:
```
# CONFIG_HELLO is not set
CONFIG_WORLD="Zephyr"
```
you will get this:
```
CONFIG_HELLO
CONFIG_WORLD
```
And nowhere in the provided scripts lines from input file are actually processed to get values, instead all comparisons of values, between files, are performed by comparisons of full lines that have been matched, in both files, to contain the same option.

I speculate that the author of the sh script has decided to use the two separate REs, because above would be really hard to achieve with use of a single basic or even a single extended regular expression, without some additional logic in sed expressions.

I think that rewriting the regular expression is OK, because it uses the features of RE that the original source code could not relay on, and the original expressions have different purpose.

mbolivar-nordic reviewed (changes requested):
Sorry it took me so long to get back to this review.

>  No it does not. Yes the logic is different, the result is the same.
>
> The outcome of this code is exactly the same dictionary object as in case of
> the `load_dot_config`, all the `is not set` options.

That is not true. Consider a line like this:

```
#                CONFIG_FOO is not set
```

With this PR, the BuildConfiguration reports CONFIG_FOO is `n`.

On master, CONFIG_FOO is not a defined configuration value.

In general this PR is mixing up refactoring with semantic changes. I am OK with moving code around and then making improvements, but this is mixing up unrelated changes that I don't think have been sufficiently justified yet.

de-nordic commented:
> In general this PR is mixing up refactoring with semantic changes. I am OK with moving code around and then making improvements, but this is mixing up unrelated changes that I don't think have been sufficiently justified yet.

Removed all my inventions that have changed semantics. And one from original `_parse`, which was the `line.strip()` thing that did not exist in `load_dot_config`.

de-nordic committed:
scripts: runners: Add BuildConfiguration.path attribute to class

The BuildConfiguration.path attribute has been added that stores path
to .config file which has been used as source for options collected to
BuildCOnfiguration object.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
scripts: runners: Copy load_dot_config to BuildConfiguration.get()

The body of load_dot_config method has been reimplemented in
BuildConfiguration.get(), replacing the previous code.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
scripts/west sign: Use BuildConfiguration instead of load_dot_config

The ImgtoolSigner.sign() is passed a BuildConfiguration object instance
that holds all of Kconfig options that have been defined for a build;
yet it has been reading the .config file once again, with the
load_dot_config, to check some of these options.
The commit replaces the code that has been using the load_dot_config
obtained data with the code that uses the BuildConfiguration object.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
scripts: west_commands: Remove no longer needed load_dot_config(path)

The functionality of the function is now covered by BuildConfiguration
class.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
@mbolivar-nordic ping

mbolivar-nordic commented on scripts/west_commands/runners/core.py:
```diff
@@ -145,30 +145,49 @@ def _init(self):
+                    _logger.warning("could not extract value from line " + line)
+                    continue
+
+                # If value not set then it is 'n'
+                if not value:
+                    value = 'n'
```
Sorry for the delay in reviewing. I'll be watching this PR closely this week.

I still think this is changing the semantics. Before, the only way I see an option `FOO` could be `n` is if the .config file says `# CONFIG_FOO is not set` (or `CONFIG_FOO=n`, which we do in our .conf fragments sometimes, though I think kconfiglib will never emit this).

But here, it looks like both `CONFIG_FOO=` and `# CONFIG_FOO=` result in `FOO` set to `n`.

Can we drop just this patch? It LGTM otherwise.

de-nordic replied:
I have reverted to the  original code as the logic to serve both matches with one regex got a little bit absurd.

The change I have added is moving the `for loop`, processing all lines, under the `with` that opens config file as did the original `_parse` code prior to bringing the `load_dot_config` body here.
I have also added `continue` at the end of positive match for `opt_val` to avoid matching the `not_set` as there is no need to check that regex against line in case when the first one matched.

de-nordic committed:
scripts: runners: Modify BuildConfiguration._parse to iterate over file

The method has been changed to iterate directly over file instead of
preloading it.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

mbolivar-nordic reviewed (approved):
Thanks for all your patience!

de-nordic commented:
> 
> 
> Thanks for all your patience!

Ha! @mbolivar  Thanks for Your patience 😄 !

de-nordic commented:
@tejlmand ping?

de-nordic commented:
@galak Can you look at this?

tejlmand commented on scripts/west_commands/runners/core.py:
```diff
@@ -144,23 +145,30 @@ def _init(self):
-            return int(value, 0)
-        except ValueError:
-            return value
+        opt_value = re.compile('^(?P<option>CONFIG_[A-Za-z0-9_]+)=(?P<value>.*)$')
```
Just a minor observation on changed behavior from existing code.

Previously, `line.strip()` was called, and all whitespaces was stripped.
This meant that a user updating `.config` by hand and by mistake introduce trailing withespaces like:
`CONFIG_TEST=y  `  would not see any problems, but with this PR, the value will now be:
`'y  '` instead of `'y'`.

I won't block the PR over this, but knowing how easy it is to accidentally introduce a stray `' '` when editing a file, and how hard it can be to trace back why the `.config` suddenly fails, I wonder if we should keep the strip.

Also the `value.endswith('"'):` further down won't match if there are trailing whitespaces.

de-nordic replied:
`.rstrip()` added.

tejlmand reviewed (commented):
Cleanup looks good.

One minor observation if @de-nordic or @mbolivar-nordic could comment.
Otherwise, i'm +1 on this.

mbolivar-nordic reviewed (approved):
Thanks @de-nordic. @tejlmand OK with you?

tejlmand reviewed (approved):
lgtm, thanks for the patience
